packages/cli/generation/ir-migrations/src/migrations/v29-to-v28/migrateFromV29ToV28.ts

```diff
@@ -0,0 +1,81 @@
+        [GeneratorName.TYPESCRIPT_BROWSER_SDK]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.TYPESCRIPT]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.TYPESCRIPT_SDK]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.TYPESCRIPT_EXPRESS]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.JAVA]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.JAVA_MODEL]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.JAVA_SDK]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.JAVA_SPRING]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.PYTHON_FASTAPI]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.PYTHON_PYDANTIC]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.OPENAPI_PYTHON_CLIENT]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.OPENAPI]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.STOPLIGHT]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.POSTMAN]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.PYTHON_SDK]: GeneratorWasNeverUpdatedToConsumeNewIR,
+        [GeneratorName.GO_FIBER]: "0.9.0-2-g6b0be0e",
+        [GeneratorName.GO_MODEL]: "0.9.0-2-g6b0be0e",
+        [GeneratorName.GO_SDK]: "0.9.0-2-g6b0be0e",
+    },
+    jsonifyEarlierVersion: (ir) =>
+        IrSerialization.V28.IntermediateRepresentation.jsonOrThrow(ir, {
+            unrecognizedObjectKeys: "strip",
+        }),
+    migrateBackwards: (v29): IrVersions.V28.ir.IntermediateRepresentation => {
+        return {
+            ...v29,
+            types: Object.fromEntries(
+                Object.keys(v29.types).map((key) => {
+                    return [key, getTypeDeclarationFromTypeId(key, v29.types)];
+                })
+            ),
+        };
+    },
+};
+
+function getTypeDeclarationFromTypeId(
+    typeId: commons.TypeId,
+    allTypes: Record<FernIr.TypeId, FernIr.TypeDeclaration>
+): FernIrV28.TypeDeclaration {
+    // Eliminate possibility of undefined
+    const typeDefinition = allTypes[typeId];
+    if (typeDefinition === undefined) {
+        throw new Error(`Type definition for type id ${typeId} is undefined`);
+    }
+
+    // Build referencedTypes
+    const newReferencedTypes: FernIrV28.DeclaredTypeName[] = [];
+    allTypes[typeId]?.referencedTypes.forEach((currentTypeId) => {
+        const currentTypeDefinition = allTypes[currentTypeId];
+        if (currentTypeDefinition === undefined) {
+            throw new Error(`Type definition for type id ${currentTypeId} is undefined`);
+        }
+        newReferencedTypes.push({
+            typeId: currentTypeId,
+            fernFilepath: currentTypeDefinition.name.fernFilepath,
+            name: currentTypeDefinition.name.name,
+        });
+    });
+
+    // Return type declaration
+    return {
+        ...typeDefinition,
+        referencedTypes: newReferencedTypes,
+    };
```

Comment:
nit: the typescript way to write this 


Some nits related to how to do this in typescript: 
```ts
function getV28TypeDeclarationFromId(
  { typeId; allTypes}: 
  {typeId: commons.TypeId, allTypes: Record<FernIr.TypeId, FernIr.TypeDeclaration>)}: FernIrV28.TypeDeclaration {
  
  const typeDeclaration = getTypeDeclarationOrThrow({ typeId, allTypes}); 
  
   return {
     ...typeDeclaration,
     referencedTypes: typeDeclaration.referencedTypes.map((referencedTypeId) => {
       const referencedTypeDeclaration = getTypeDeclarationOrThrow({ typeId: referencedTypeId, allTypes),
       return {
         typeId: referencedTypeId,
         fernFilePath: referencedTypeDeclaration.name.fernFilepath,
         name: referencedTypeDeclaration.name.name,
       }
     }),
   }
}

function getTypeDeclarationOrThrow(
  { typeId; allTypes}: 
  {typeId: commons.TypeId, allTypes: Record<FernIr.TypeId, FernIr.TypeDeclaration>):  FernIr.TypeDeclaration {
  
  }
```

Reply:
boils down to: 
- constructing immutable objects using `map`
- having a `getTypeDeclarationOrThrow` will make it easier